src/rest_vol_dataset.c

```diff
@@ -5023,86 +5106,155 @@ RV_json_values_to_binary_callback(char *HTTP_response, const void *callback_data
-                FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_PARSEERROR, FAIL, "failed to parse member datatype");
+                if ((member_val = YAJL_GET_OBJECT(value_entry)->values[i]) == NULL)
+                    FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_PARSEERROR, FAIL,
+                                    "failed to parse member of compound type");
 
-            offset += member_size;
-            member_val      = NULL;
-            member_size     = 0;
-            member_dtype_id = H5I_INVALID_HID;
-        }
-    }
-    else {
-        FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL, "unsupported datatype class for parsing");
+                if (RV_json_values_to_binary_recursive(member_val, member_dtype_id,
+                                                       (void *)((char *)value_buffer + offset)) < 0)
+                    FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_PARSEERROR, FAIL, "failed to parse member datatype");
+
+                offset += member_size;
+                member_val      = NULL;
+                member_size     = 0;
+                member_dtype_id = H5I_INVALID_HID;
+            }
+        } break;
+
+        case (H5T_VLEN): {
+            yajl_val seq_elem_val      = NULL;
+            size_t   parent_dtype_size = 0;
+            void    *seq_elem_ptr      = NULL;
+            vl                         = (hvl_t *)value_buffer;
+
+            if ((parent_dtype_id = H5Tget_super(dtype_id)) == H5I_INVALID_HID)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get base type of vlen sequence");
+
+            if ((parent_dtype_size = H5Tget_size(parent_dtype_id)) == 0)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get size of vlen parent type");
+
+            if (!YAJL_IS_ARRAY(value_entry))
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_PARSEERROR, FAIL,
+                                "server returned invalid json for vlen sequence");
+
+            if (YAJL_GET_ARRAY(value_entry) == NULL)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_PARSEERROR, FAIL, "couldn't parse returned vlen sequence");
+
+            if ((vl->len = YAJL_GET_ARRAY(value_entry)->len) <= 0)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_PARSEERROR, FAIL,
+                                "server returned vlen sequence with invalid length");
+
+            if ((vl->p = calloc(vl->len, parent_dtype_size)) == NULL)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL,
+                                "can't allocate space for returned vlen seq data");
+
+            /* Recursively encode each element in the sequence and place it in the allocated buffer */
+            for (size_t i = 0; i < vl->len; i++) {
+                if ((seq_elem_val = YAJL_GET_ARRAY(value_entry)->values[i]) == NULL)
+                    FUNC_GOTO_ERROR(H5E_DATASET, H5E_PARSEERROR, FAIL, "can't parse vlen sequence element");
+
+                seq_elem_ptr = (void *)(((char *)vl->p) + i * parent_dtype_size);
```

Comment:
Fine for now, though it seems this would only work for simple cases. For example, if the parent datatype is also a vlen I would expect the calculations here for the next element to not be correct. The function would likely need to be able to pass the current offset into the buffer both down and back up.